File: algorithms/binary_search/reverse_pairs_merge_sort.py
# https://leetcode.com/problems/reverse-pairs/
# https://leetcode.com/problems/reverse-pairs/discuss/97287/C%2B%2B-with-iterators

# time complexity: O(nlog(n))

class Solution:
    def reversePairs(self, nums: list[int]) -> int:
        def merge_sort(begin: int, end: int) -> int:
            if end - begin < 2:
                return 0
            right = mid = (begin + end) // 2
            cnt = merge_sort(begin, mid) + merge_sort(mid, end)
            for left in range(begin, mid):
                while right < end and nums[left] > 2 * nums[right]:
                    right += 1
                cnt += right - mid
            nums[begin:end] = sorted(nums[begin:end])
            return cnt
        return merge_sort(0, len(nums))

File: algorithms/binary_search/reverse_pairs_merge_sort.cc
// https://leetcode.com/problems/reverse-pairs/
// https://leetcode.com/problems/reverse-pairs/discuss/97287/C%2B%2B-with-iterators

// time complexity: O(nlog(n))

#include <vector>

using namespace std;

class Solution {
    int _len;

public:
    int merge_sort(vector<int>::iterator begin, vector<int>::iterator end) {
        _len = end - begin;
        if (_len < 2)
            return 0;

        auto mid = begin + _len / 2;
        int cnt = merge_sort(begin, mid) + merge_sort(mid, end);
        for (auto left = begin, right = mid; left < mid; left++) {
            while (right < end && (*left > 2L * *right))
                right++;
            cnt += right - mid;
        }
        inplace_merge(begin, mid, end);
        return cnt;
    }
    int reversePairs(vector<int> &nums) {
        return merge_sort(nums.begin(), nums.end());
    }
};